manubot/cite/citekey.py

```diff
@@ -1,82 +1,171 @@
-"""Functions importable from manubot.cite submodule (submodule API):
-
-  standardize_citekey()
-  citekey_to_csl_item()
-
-Helpers:
-
-  inspect_citekey()
-  is_valid_citekey() - also used in manubot.process
-  shorten_citekey() - used solely in manubot.process
-  infer_citekey_prefix()
-
+"""
+Utilities for representing and processing citation keys.
 """
 import functools
 import logging
 import re
+import dataclasses
 
-from manubot.util import import_function
+try:
+    from functools import cached_property
+except ImportError:
+    from backports.cached_property import cached_property
 
-citeproc_retrievers = {
-    "doi": "manubot.cite.doi.get_doi_csl_item",
-    "pmid": "manubot.cite.pubmed.get_pubmed_csl_item",
-    "pmcid": "manubot.cite.pubmed.get_pmc_csl_item",
-    "arxiv": "manubot.cite.arxiv.get_arxiv_csl_item",
-    "isbn": "manubot.cite.isbn.get_isbn_csl_item",
-    "wikidata": "manubot.cite.wikidata.get_wikidata_csl_item",
-    "url": "manubot.cite.url.get_url_csl_item",
-}
 
-"""
-Regex to extract citation keys.
-The leading '@' is omitted from the single match group.
+@dataclasses.dataclass
+class CiteKey:
+    input_id: str
+    aliases: dict = dataclasses.field(default_factory=dict)
 
-Same rules as pandoc, except more permissive in the following ways:
+    def __post_init__(self):
+        self.check_input_id(self.input_id)
 
-1. the final character can be a slash because many URLs end in a slash.
-2. underscores are allowed in internal characters because URLs, DOIs, and
-   citation tags often contain underscores.
+    @staticmethod
+    def check_input_id(input_id):
+        if not isinstance(input_id, str):
+            raise TypeError(
+                "input_id should be type 'str' not "
+                f"{type(input_id).__name__!r}: {input_id!r}"
+            )
+        if input_id.startswith("@"):
+            f"invalid citekey input_id: {input_id!r}\nstarts with '@'"
 
-If a citekey does not match this regex, it can be substituted for a
-tag that does, as defined in citation-tags.tsv.
+    @classmethod
+    @functools.lru_cache(maxsize=None)
+    def from_input_id(cls, *args, **kwargs):
```

Comment:
Where is this used?

Reply:
nowhere yet, but I thought it'd be good to have a way to not make duplicate objects for the same input_id.